### ISSUE TITLE
Duplicate icons in notification area

### ISSUE THREAD

**Issue by hoz816** [mate-desktop/mate-panel#539]
Mate Panel 1.18 on Arch Linux.  An undersized icon is displayed in the notification area when the clipboard manager (clipit) is first started.  When a clipboard function is first accessed (on copy for instance), a second, larger version of the icon is displayed next to it.  This behavior was not present prior to upgrading. 
![screenshot at 2017-03-19 20-46-21](https://cloud.githubusercontent.com/assets/26531062/24086401/26725988-0ce5-11e7-964d-417d4b7ef1a3.png)
![screenshot at 2017-03-19 20-45-39](https://cloud.githubusercontent.com/assets/26531062/24086402/2678b51c-0ce5-11e7-9152-441ebba24e96.png)

**Comment by hoz816** [comment]
This behavior seem to be isolated to this specific clipboard manager and it's libappindicator dependency (both GTK2 and GTK3 versions).  

Replaced clipit with parcellite, and all is well.

**Comment by cwendling** [comment]
> An undersized icon is displayed in the notification area when the clipboard manager (clipit) is first started.

This looks like it is a bug in the theme, but it should have been fixed already in https://github.com/mate-desktop/mate-themes/commit/b64a6c90a982a67a4a74d38dc60c6f0b57d3478d, yet Arch still has them:

```console
$ grep -rF NaTrayApplet-icon-size /usr/share/themes/
/usr/share/themes/ContrastHighInverse/gtk-3.0/mate-applications.css:    -NaTrayApplet-icon-size: 16;
/usr/share/themes/TraditionalOk/gtk-3.0/mate-applications.css:    -NaTrayApplet-icon-size: 16;
/usr/share/themes/GreenLaguna/gtk-3.0/mate-applications.css:    -NaTrayApplet-icon-size: 16;
/usr/share/themes/BlackMATE/gtk-3.0/mate-applications.css:    -NaTrayApplet-icon-size: 16;
/usr/share/themes/Green-Submarine/gtk-3.0/mate-applications.css:    -NaTrayApplet-icon-size: 16;
/usr/share/themes/Blue-Submarine/gtk-3.0/mate-applications.css:    -NaTrayApplet-icon-size: 16;
/usr/share/themes/BlueMenta/gtk-3.0/mate-applications.css:    -NaTrayApplet-icon-size: 16;
/usr/share/themes/TraditionalGreen/gtk-3.0/mate-applications.css:    -NaTrayApplet-icon-size: 16;
/usr/share/themes/Menta/gtk-3.0/mate-applications.css:    -NaTrayApplet-icon-size: 16;
```

> When a clipboard function is first accessed (on copy for instance), a second, larger version of the icon is displayed next to it.

This looks like either clipit or libappindicator is buggy: when doing a clipboard action, clipit tries to register the same icon again, which fails (expectedly, only 1 item with the very same properties can exist on the bus), and I guess it then fallbacks to an traditional tray icon.

If watching the *org.kde.StatusNotifierWatcher* interface, we can see the incorrect registration calls (not followed by the corresponding *registered* signal):
```console
$ dbus-monitor --session --monitor interface=org.kde.StatusNotifierWatcher
[…]
method call time=1490204004.696659 sender=:1.40 -> destination=:1.33 serial=7 path=/StatusNotifierWatcher; interface=org.kde.StatusNotifierWatcher; member=RegisterStatusNotifierItem
   string "/org/ayatana/NotificationItem/clipit"
signal time=1490204004.696857 sender=:1.33 -> destination=(null destination) serial=35 path=/StatusNotifierWatcher; interface=org.kde.StatusNotifierWatcher; member=StatusNotifierItemRegistered
   string ":1.40/org/ayatana/NotificationItem/clipit"
method call time=1490204005.185137 sender=:1.40 -> destination=:1.33 serial=12 path=/StatusNotifierWatcher; interface=org.kde.StatusNotifierWatcher; member=RegisterStatusNotifierItem
   string "/org/ayatana/NotificationItem/clipit"
method call time=1490204026.056138 sender=:1.40 -> destination=:1.33 serial=21 path=/StatusNotifierWatcher; interface=org.kde.StatusNotifierWatcher; member=RegisterStatusNotifierItem
   string "/org/ayatana/NotificationItem/clipit"
method call time=1490204028.566168 sender=:1.40 -> destination=:1.33 serial=25 path=/StatusNotifierWatcher; interface=org.kde.StatusNotifierWatcher; member=RegisterStatusNotifierItem
   string "/org/ayatana/NotificationItem/clipit"
```
the 2 first blocks are legitimate (first registration, and then ACK from the service), but the subsequent ones shouldn't happen.

**Comment by monsta** [comment]
Thanks @cwendling, now I'm pretty sure the similar issue [with nm-applet](https://ubuntu-mate.community/t/17-04-nm-applet-works-fine-but-shows-2-icons-in-the-system-tray/12020) is caused by nm-applet itself... :slightly_smiling_face: 

@hoz816: I see you don't have the issue with nm-applet in Arch. What's the version of nm-applet there?

**Comment by hoz816** [comment]
Currently running version 1.4.6 of the nm-applet.  So far, clipit has been the only issue that I've had with duplicate icons.

**Comment by monsta** [comment]
Ok, so might be a problem of earlier nm-applet versions.

**Comment by cwendling** [comment]
@monsta I see Ubuntu's nm-applet is built with appindicator support, and they enable it by default, while Debian's isn't.  I'd guess that's the reason, and it's the same issue than with clipit, suggesting the issue lies in libappindicator.

See
* http://packages.ubuntu.com/zesty/network-manager-gnome
* https://git.launchpad.net/~network-manager/network-manager/+git/ubuntu-applet/tree/debian/patches/Have-the-appindicator-enabled-by-default.patch
* https://packages.debian.org/sid/network-manager-gnome

**Comment by rilian-la-te** [comment]
I think there is MATE proto implementation problem.
When I wrote xfce4-sntray-plugin (It also works in mate, regardless of name), I consider than reregistration is legal (because I see exactly same issue).
But AFAIK, MATE StatusNotifier does not allow reregistration. But libappindicator used this ability.

**Comment by cwendling** [comment]
@rilian-la-te Thanks a lot for the info.  Do you know if there is any actual reason for re-registration?  I don't see why that would be useful, but I might miss it.
Also, what do you do then? drop the old one and register the new one, or take it as a mere update?  It's probably easier to de-register and re-register, but might flicker I guess.

I'll take a look at the spec to see if it talks about it, but likely see to handle it anyway if it's indeed what appindicator always does. Sad.

FWIW, the SNI implementation comes from gnome-flashback, so if it's indeed an implementation problem they'll have the same issue.  When we fix this should propose it there.

**Comment by rilian-la-te** [comment]
Here is my implementation of registration:
`        public void register_status_notifier_item(string service, BusName sender)
        {
            var is_path = (service[0]=='/') ? true : false;
            string path, name;
            if (is_path)
            {
                name = (string)sender;
                path = service;
            }
            else
            {
                name = service;
                path = "/StatusNotifierItem";
            }
            var id = get_id(name,path);
            if (id in name_watcher)
            {
                warning("Trying to register already registered item. Reregistering new...");
                remove(id);
            }
            var name_handler = Bus.watch_name(BusType.SESSION,name,GLib.BusNameWatcherFlags.NONE,
                                                ()=>{
                                                    try {
                                                        ItemIface ping_iface = Bus.get_proxy_sync(BusType.SESSION,name,path);
                                                        ping_iface.notify.connect((pspec)=>{
                                                            if (ping_iface.id == null ||
                                                            ping_iface.title == null ||
                                                            ping_iface.id.length <= 0 ||
                                                            ping_iface.title.length <= 0)
                                                                remove(get_id(name,path));
                                                        });
                                                    } catch (Error e) {remove(get_id(name,path));}
                                                },
                                                () => {remove(get_id(name,path));}
                                                );
            name_watcher.insert(id,name_handler);
            status_notifier_item_registered(id);
            this.notify_property("registered-status-notifier-items");
        }`
I does not know re-registration reason, but Unity used it a lot (because my code comes from Unity's indicator plugin).

**Comment by monsta** [comment]
> When I wrote xfce4-sntray-plugin (It also works in mate, regardless of name), I consider than reregistration is legal (because I see exactly same issue).

I wonder if Xubuntu uses it. In https://bugs.launchpad.net/bugs/1652454 it was mentioned that there's the same duplicate issue.

**Comment by monsta** [comment]
@cwendling: thanks, https://github.com/mate-desktop/mate-panel/commit/2f5fb93bb06fd13e60d18b7b0bfcf0a0b13f2a47 fixed it for me in Mint 18.1 (based on Ubuntu 16.04) :slightly_smiling_face:
